### PR TITLE
Karma Init

### PR DESCRIPTION
**Pull request by sauravkhdoolia**
| Q                | A
| ---------------- | ---
| Bug fix?         | yes
| New feature?     | no
| BC breaks?       | no
| Deprecations?    | no
| New tests added? | not needed
| License          | MIT

### Description

- [x] Moving to Karma for handling tests.
- [x] Moved to BrowserStack from SauceLabs
- [x] Coveralls Working
- [x] Tests passing on Browsers
    - [x] Win 10 Chrome
    - [x] Win 10 Firefox
    - [x] Win 10 Edge
    - [x] Win 8.1 Chrome
    - [x] Win 8.1 Edge
    - [x] OSX El Capitan Chrome
    - [x] OSX El Capitan Firefox
    - [x] OSX El Capitan Safari

--

### CODE REVIEW

**Comment by sauravkhdoolia** [comment]
#1552 

Tests passing again!
Browser-Stack being utilized.

**Comment by sauravkhdoolia** [comment]
`
Firefox 78.0 (Windows 10): Executed 617 of 617 (5 FAILED) (18.184 secs / 10.708 secs)
TOTAL: 5 FAILED, 612 SUCCESS


1) fontSize support old style
     Core-API execAction
     TypeError: MediumEditor.util.getContainerEditorElement(...).focus is not a function in src/js/core.js (line 962)
execAction@src/js/core.js:962:94
@spec/core-api.spec.js:239:20
<Jasmine>

2) fontName support new stle
     Core-API execAction
     TypeError: MediumEditor.util.getContainerEditorElement(...).focus is not a function in src/js/core.js (line 962)
execAction@src/js/core.js:962:94
@spec/core-api.spec.js:225:20
<Jasmine>

3) fontSize support new stle
     Core-API execAction
     TypeError: MediumEditor.util.getContainerEditorElement(...).focus is not a function in src/js/core.js (line 962)
execAction@src/js/core.js:962:94
@spec/core-api.spec.js:253:20
<Jasmine>

4) fontName support old style
     Core-API execAction
     TypeError: MediumEditor.util.getContainerEditorElement(...).focus is not a function in src/js/core.js (line 962)
execAction@src/js/core.js:962:94
@spec/core-api.spec.js:211:20
<Jasmine>

5) should pass opt directly to document.execCommand
     Core-API execAction
     TypeError: MediumEditor.util.getContainerEditorElement(...).focus is not a function in src/js/core.js (line 962)
execAction@src/js/core.js:962:94
@spec/core-api.spec.js:197:20
<Jasmine>


TOTAL: 5 FAILED, 612 SUCCESS
`

**Comment by sauravkhdoolia** [comment]
```
Safari 8.0.8 (Mac OS 10.10.5): Executed 617 of 617 (4 FAILED) (3.991 secs / 3.631 secs)
TOTAL: 4 FAILED, 613 SUCCESS


1) should use custom placeholder text when passed as the placeholder.text option
     MediumEditor.extensions.placeholder TestCase
     Expected '' to match /^['"]Custom placeholder['"]$/.
<Jasmine>
validatePlaceholderContent@spec/placeholder.spec.js:152:40
spec/placeholder.spec.js:219:35
<Jasmine>

2) should use the data-placeholder when it is present
     MediumEditor.extensions.placeholder TestCase
     Expected '' to match /^['"]Custom placeholder['"]$/.
<Jasmine>
validatePlaceholderContent@spec/placeholder.spec.js:152:40
spec/placeholder.spec.js:209:35
<Jasmine>

3) should add the default placeholder text when data-placeholder is not present on dynamically added elements
     MediumEditor.extensions.placeholder TestCase
     Expected '' to match /^['"]Type your text['"]$/.
<Jasmine>
validatePlaceholderContent@spec/placeholder.spec.js:152:40
spec/placeholder.spec.js:168:35
<Jasmine>

4) should add the default placeholder text when data-placeholder is not present
     MediumEditor.extensions.placeholder TestCase
     Expected '' to match /^['"]Type your text['"]$/.
<Jasmine>
validatePlaceholderContent@spec/placeholder.spec.js:152:40
spec/placeholder.spec.js:159:35
<Jasmine>
```

File: src/js/core.js [hunk 1]
```diff
@@ -957,8 +957,8 @@
             }
 
             // https://github.com/yabwe/medium-editor/issues/1496
-            // ensure the focus remains in the editor for Firefox
-            if (MediumEditor.util.isFF) {
+            // But only if something is selected i.e, getSelectedParentElement !== {}
+            if (this.getSelectedParentElement().length > 0 && MediumEditor.util.isFF) {
```
Comment (sauravkhdoolia) [inline src/js/core.js:961]: Root cause, In some cases when `execCommand` was called in FF, it checked for Firefox but never checked if an element is selected or not! And Hence, the error 
```MediumEditor.util.getContainerEditorElement(this.getSelectedParentElement()).focus()``` .

Resolved.

**Comment by sauravkhdoolia** [comment]
IE is used by <2% users in the world. I think we should just drop it.

[Browser Share](https://www.statista.com/statistics/544400/market-share-of-internet-browsers-desktop/)

**Comment by arteconceito** [comment]
> IE is used by <2% users in the world. I think we should just drop it.
> @roniemartinez What you think about IE support!
> 
> [Browser Share](https://www.statista.com/statistics/544400/market-share-of-internet-browsers-desktop/)

In general public usage yes but believe me, there are still many companies (I mean at the enterpise level) that still and only use IE 11.
I suppose this should be more a question of... **Do we want to support it or not?**

By the way, nice work with the tests migration. Unfortunately, I could not yet find the time I wished to contribute to the project!

**Comment by sauravkhdoolia** [comment]
Yeah, I think the same too.

I think we should go for v10 and v11 as most people uses ME cause it used to support all browsers.

Yeah, I got some free time(more like spirit than time) to do this. As i barely write tests.

And I'm still clueless on debugging these tests on IE and supporting them.

@arteconceito  Direct me? I'll try to do it.

**Comment by arteconceito** [comment]
@sauravkhdoolia in my opinion supporting IE v11 is good enough and should cover most of IE user base as of today.

In order to debug IE I usually use a VM running IE11 on it. Firebug is also a good debugging extension as IE dev tools are not the best.

**Comment by sauravkhdoolia** [comment]
I'll start there. Thanks brother.

**Comment by sauravkhdoolia** [comment]
Hey @arteconceito It works on IE11 sometimes and sometimes it doesn't.

That is so weird to catch, Why sometimes an API is available in the IE browser, and sometimes it's not.

Here:

**Comment by sauravkhdoolia** [comment]
@arteconceito It works in IE11, but sometimes it doesn't.

I don't know why?

Here: 

Medium Edittor IE test screenshot

**Comment by sauravkhdoolia** [comment]
IE Errors catch on BrowserStack.

```
IE 11.0 (Windows 10): Executed 617 of 617 (2 FAILED) (20.473 secs / 12.935 secs)
TOTAL: 2 FAILED, 615 SUCCESS


1) should change font name when select is changed
     Font Name Button TestCase Font Name
     Error: <toHaveBeenCalledWith> : Expected a spy, but got Function.
Usage: expect(<spyObj>).toHaveBeenCalledWith(...arguments)
Error: <toHaveBeenCalledWith> : Expected a spy, but got Function.
Usage: expect(<spyObj>).toHaveBeenCalledWith(...arguments)
    at <Jasmine>
   at Anonymous function (spec/fontname.spec.js:70:13)
    at <Jasmine>

2) should revert font name when select value is set to empty
     Font Name Button TestCase Font Name
     Error: <toHaveBeenCalledWith> : Expected a spy, but got Function.
Usage: expect(<spyObj>).toHaveBeenCalledWith(...arguments)
Error: <toHaveBeenCalledWith> : Expected a spy, but got Function.
Usage: expect(<spyObj>).toHaveBeenCalledWith(...arguments)
    at <Jasmine>
   at Anonymous function (spec/fontname.spec.js:104:13)
    at <Jasmine>


TOTAL: 2 FAILED, 615 SUCCESS
```

**Pull request merged**